tools/wptrunner/wptrunner/wptlogging.py

```diff
@@ -133,3 +133,4 @@ def __exit__(self, *args, **kwargs):
                         pass
                 self.logging_queue.close()
                 self.logger.info("queue closed")
+                self.logger.shutdown()
```

Comment:
So, I don't think this is going to work. Sometimes we (hackily?) run multiple session with the same logger (e.g. for stability checking). I think you need to look for all the places in the code where we call `setup_logging` and ensure that each of those places has a corresponding call to `logger.shutdown()` once the logger is no longer in use.

Reply:
But don't you think that if a command, like mentioned in original issue, is going to fail, then the control will never return to the place, where I have called `setup_logging`, so even if I define the shutdown function at each place, the logger will not shut down properly and nothing will be written to the log file. The command fails at the environment variable initialization part in the `wptrunner.py` file, and that was the reason that the `suite_end()` function was never called. Also, I could not understand what you want to say when you mention about _run multiple session with the same logger_? If you could provide a command which gives such scenario, that would be a great help for testing.

Reply:
I think that you want to change the call in `wptlogging.py` for successful calls only. When the tests fail, the call in `wptrunner.py` will call the `shutdown()` method and write to the file correctly. But when the test cases succeed, the control will reach the end, and hence the `shutdown()` calls will also execute. If this is what you want me to do, then I think, I agree with you here. I'll do the necessary changes. If not, please elaborate your opinion about the problem.

Reply:
So when there's an exception the stack will unwind, but we'll still have the oppertunty to execute exception handlers and finally cluases in each calling function.

The way that the logger is used isn't ideal because we tend to call `setup_logging` and set a global, but it's quite possible to ensure that we have a `try/finally` block enclosing the scope over which the logger is actually used, and calling shotdown if we exit that scope even via an exception.

Reply:
Hi, sorry for the delay. I have updated the code to wrap all the `setup_logging` calls in `try/finally` blocks. I have kept my previous logic of calling `shutdown()` in `wptlogging.py`'s `CaptureIO` class as without that nothing was being written to the `empty.json` file. Please review and give suggestions.